prereise/gather/hydrodata/eia/fetch_historical.py

```diff
@@ -0,0 +1,129 @@
+    }
+    available_regions = region_categories.keys()
+    if regions is None:
+        regions = region_categories.keys()
+    parsed_regions = parse_input_regions(available_regions, regions)
+    parsed_series_data = []
+    skipped_regions = set()
+    for region in tqdm(parsed_regions, total=len(parsed_regions)):
+        # This request gives us the available sub-sub-categories (fuel, timezone)
+        url = category_template.format(
+            api_key=api_key, category_id=region_categories[region]
+        )
+        subcategories_response = requests.get(url).json()
+        # Filter these by their names to hydro, UTC
+        matching_series_set = {
+            series["series_id"]
+            for series in subcategories_response["category"]["childseries"]
+            if "hourly - UTC time" in series["name"] and "hydro" in series["name"]
+        }
+        # Check for existence of UTC hydro data
+        if len(matching_series_set) == 0:
+            skipped_regions.add(region)
+            continue
+        # Extract the only value from a singleton set
+        (series_id,) = matching_series_set
+        # Query just that series
+        series_response = requests.get(
+            series_template.format(api_key=api_key, series_id=series_id)
+        ).json()
+        # Parse the full response to a Series containing just the data
+        series_data = (
+            pd.DataFrame(series_response["series"][0]["data"]).set_index(0).squeeze()
+        )
+        series_data.index = pd.to_datetime(series_data.index)
+        series_data.name = region
+        parsed_series_data.append(series_data)
+    if len(skipped_regions) > 0:
+        print(f"no hydro data found for {sorted(skipped_regions)}")
+    full_data = pd.concat(parsed_series_data, axis=1)
+    # Evaluate the number of missing entries for each region
+    num_missing = full_data.loc[f"{year}-01-01-00":f"{year}-12-31-23"].isna().sum()
+    acceptable_num_missing = num_missing <= acceptable_missing
+    imputed_regions = full_data.columns[
+        (0 < num_missing) & (num_missing <= acceptable_missing)
+    ]
+    missing_entry_regions = full_data.columns[~acceptable_num_missing]
+    full_year_data_regions = full_data.columns[acceptable_num_missing]
+    if len(imputed_regions) > 0:
+        print(f"acceptable missing entries for {sorted(imputed_regions)}, imputing")
+    if len(missing_entry_regions) > 0:
+        print(f"inadequate data for {sorted(missing_entry_regions)} for {year}")
+    if verbose and len(missing_entry_regions) > 0:
+        print(num_missing.loc[missing_entry_regions])
+    # Impute
+    imputed = linear(
+        full_data.loc[f"{year}-01-01-00":f"{year}-12-31-23", full_year_data_regions],
+        inplace=False,
+    )
+    # Rename columns for brevity. 'Longname (ABV)' becomes simply 'ABV'
+    imputed.columns = [
+        re.search(r"\(([A-Z]*[0-9]*)\)", c).group(1) for c in imputed.columns
+    ]
+    return imputed
```

Comment:
Do you think this function can be generalize to any fuel type? We do similar things in `prereise.gather.helpers.py` using EIA form 923.

Reply:
It could be done, assuming the same API patterns from EIA, and using a dictionary for `top_level_category_id` instead of a single fixed value (which is hydro-specific). What would we use the resulting data for, validation?

Reply:
This is where the `get_monthly_net_generation` from the `helpers` module is used:
```
prereise/gather/tests/test_helpers.py:from prereise.gather.helpers import get_monthly_net_generation
prereise/gather/tests/test_helpers.py:def test_get_monthly_net_generation_argument_type():
prereise/gather/tests/test_helpers.py:            get_monthly_net_generation(a[0], a[1], a[2])
prereise/gather/tests/test_helpers.py:def test_get_monthly_net_generation_argument_value():
prereise/gather/tests/test_helpers.py:            get_monthly_net_generation(a[0], a[1], a[2])
prereise/gather/tests/test_helpers.py:def test_get_monthly_net_generation():
prereise/gather/tests/test_helpers.py:        get_monthly_net_generation(state, table, fuel_type) for fuel_type in fuel_types
prereise/gather/hydrodata/eia/demo/eastern_hydro_v3_demo/eastern_hydro_v3_demo.ipynb:    "                                     get_monthly_net_generation)\n",
prereise/gather/hydrodata/eia/demo/eastern_hydro_v3_demo/eastern_hydro_v3_demo.ipynb:    "        eastern_monthly_hyc_rest[state] = get_monthly_net_generation(state, eia_923_form, 'hydro', hps=False)\n",
prereise/gather/hydrodata/eia/demo/eastern_hydro_v3_demo/eastern_hydro_v3_demo.ipynb:    "        total_state_profile = get_monthly_net_generation(state, eia_923_form, 'hydro', hps=False)\n",
prereise/gather/hydrodata/eia/demo/eastern_hydro_v3_demo/eastern_hydro_v3_demo.ipynb:    "        total_state_profile = get_monthly_net_generation(state, eia_923_form, 'hydro', hps=False)\n",
prereise/gather/hydrodata/eia/demo/western_hydro_v2_demo.ipynb:    "from prereise.gather.helpers import get_monthly_net_generation\n",
prereise/gather/hydrodata/eia/demo/western_hydro_v2_demo.ipynb:    "hydro_monthly_generation_ca = get_monthly_net_generation('CA', plant_generation, 'hydro')"
prereise/gather/hydrodata/eia/demo/western_hydro_v2_demo.ipynb:    "hydro_monthly_generation_wa = get_monthly_net_generation('WA', plant_generation, 'hydro')"
prereise/gather/hydrodata/eia/demo/western_hydro_v2_demo.ipynb:    "    hydro_monthly_generation = get_monthly_net_generation(s, plant_generation, 'hydro')\n",
prereise/gather/hydrodata/eia/demo/western_hydro_v2_demo.ipynb:    "hydro_monthly_generation_wy = get_monthly_net_generation('WY', plant_generation, 'hydro')\n",
"hydro_monthly_generation_wy = get_monthly_net_generation('WY', plant_generation, 'hydro')\n",
```
So it used to generate TAMU hydro profiles.

Reply:
It does not need to be done here but we should keep in mind that we can use the API to get the generation and avoid using EIA orm 963. Do they also have an API for the capacity of the generators (form 860)?

Reply:
We've done some thinking about this in the past: https://github.com/Breakthrough-Energy/PreREISE/issues/246

There are potentially several different sources of Form 860 data.

Reply:
I am glad we have this API now. We should try to take full advantage of it in the package but this is out of the scope of this PR.